haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -27,42 +397,280 @@ import qualified Concordium.Crypto.SHA256 as Hash
 import Concordium.ID.Parameters
 import Concordium.Types
 import Concordium.Types.HashableTo
-import Concordium.Utils
 
 -- |Chain cryptographic parameters.
 type CryptographicParameters = GlobalContext
 
--- |The mint-per-slot rate is part of the 'MintDistribution' parameters for 'ChainParametersV0',
--- but as of 'ChainParametersV1' is replaced by a mint-per-payday rate in the 'TimeParameters'.
--- 'MintPerSlotForCPV0'
-data MintPerSlotForCPV0 cpv where
-    MintPerSlotForCPV0Some :: {_mpsMintPerSlot :: !MintRate} -> MintPerSlotForCPV0 'ChainParametersV0
-    MintPerSlotForCPV0None :: MintPerSlotForCPV0 'ChainParametersV1
-
-instance IsChainParametersVersion cpv => Serialize (MintPerSlotForCPV0 cpv) where
-    put MintPerSlotForCPV0Some{..} = put _mpsMintPerSlot
-    put MintPerSlotForCPV0None = return ()
-    get = case chainParametersVersion @cpv of
-        SCPV0 -> MintPerSlotForCPV0Some <$> get
-        SCPV1 -> return MintPerSlotForCPV0None
-
--- |Lens for '_mpsMintPerSlot'
-{-# INLINE mpsMintPerSlot #-}
-mpsMintPerSlot :: Lens' (MintPerSlotForCPV0 'ChainParametersV0) MintRate
-mpsMintPerSlot =
-    lens _mpsMintPerSlot (\mps x -> mps{_mpsMintPerSlot = x})
-
-deriving instance Eq (MintPerSlotForCPV0 cpv)
-deriving instance Show (MintPerSlotForCPV0 cpv)
+$( singletons
+    [d|
+        data MintDistributionVersion
+            = MintDistributionVersion0
+            | MintDistributionVersion1
+
+        mintDistributionVersionFor :: ChainParametersVersion -> MintDistributionVersion
+        mintDistributionVersionFor ChainParametersV0 = MintDistributionVersion0
+        mintDistributionVersionFor ChainParametersV1 = MintDistributionVersion1
+        mintDistributionVersionFor ChainParametersV2 = MintDistributionVersion1
+
+        supportsMintPerSlot :: MintDistributionVersion -> Bool
```

Comment:
That is kind of what we did before. The problem with that approach is that (with the new introduced versions, parameter types etc.) it would become quite a lot of boiler plate code.

Reply:
Just to be clear, my comment is not about using the singleton library, but more of using template haskell for generating this part, instead of doing it manually and still using the library.
I don't have a good grasp of how much work and whether it would be worth it

Reply:
Resolving this. I will add more documentation for the splice and what it generates. There's also another comment pointing out this issue.